File: merge_sort.py
# Abdul's python implementation
        
        def merge(a , b):
          if(len(a) == 0 and len(b) ==0)
              return None
           else if(len(a) == 0 and len(b) > 0 )
             return b 
           else if (len(b) == 0 and len(a) > 0 )
             return a
           else 
          		index1 = 0
            	index2 = 0 
                combined = [] 
             	while index1 < len(a) and index2 < len(b):
                  if a[index1] >= b[index2]:
                  	 combined.append(a[index1])
                    index1 += 1;	
             	  else 
             		combined.append(b[index2])
                    index2 +=1
            # exhaust b   
             if(index1 != len(a)) 
                  for i in range(index2, len(b)) 
                    combined.append(b[i])
             # exhaust a
             if(index2 != len(b)):
                 for i in range(index1, len(a)) 
                    combined.append(a[i])
             return combined
        
        
       def mergesort(int left, int right ,int a[]):
          if(len(a) <= 1)
            return a; 
          
          mid = len(a) / 2 
          
          left_arr = mergesort(left , mid , a[left:mid]) 
          right_arr = mergesort(mid , right , a[mid:right+1])
          
          return merge(left_arr, right_arr)
          

File: merge_sort.c
  /*Merge Sort*/
  
  
  
  
  void merge(int* array, int left_ind, int mid, int right_ind){
	 int left_size = (mid - left_ind + 1);
  	 int right_size = (right_ind - mid + 1);
  	 
  	 int left[left_size], right[right_size];
     for(int i = 0; i < left_size; i++){
  		left[i] = array[i];
     }
  	 for(int j = 0; j < right_size; j++){
  	    right[j] = array[mid+i];
     }
  	 int x = 0, y = 0;
  	 for (int k=left_index; k < right_ind; k++){
       	if (x < left.length && left[x] <= right[y]){
          array[k] = left[x];
          x++;
        } else if(y < right.length && left[x] > right[y]) {
          array[k] = right[y];
          y++;
        } else if (x < left_size){
          array[k] = left[x];
        } else if (y < right_size){
          array[k] = right[x];	
        }
       	k++;
     }
    
     
  
     
  }
  
  void merge_sort(int* array, int left_ind, int right_ind){
  	 
     int mid = (right_ind - left_ind)/2;
     if (left_ind < right_ind){
       merge_sort(array, left_ind, mid);
  	   merge_sort(array, mid+1, right_ind);
      
       merge(array, left_ind, mid, right_ind);
     }
  }
  
  
  
  
  
    
  
        
        
        
 /*       
  
  
  Find the kth smallest value in unsorted array
  
  
  brute force 
  sort it lol 
  heapify the array and pop the top k times 
  klog(n)
  Merge sort + k times searcg 
  knlog(n)
  
  2nd smallest 
  
  = 3 
  
  4 th smallest
  
  length = 5 
  
  
  5 3 7 8 2 
  
  
  */
        
        
        
        
        